### PR TITLE
VULCAN-21 Add Vulcan Documentation

### PR DESCRIPTION
**Pull request by ahdezma**

### CODE REVIEW

File: vulcan-api/src/main/java/com/liferay/vulcan/message/json/PageMessageMapper.java [hunk 1]
```diff
@@ -21,121 +21,315 @@
...
+	 * @return the media type for this mapper.
+	 */
 	public String getMediaType();
 
+	/**
+	 * Maps a collection URL to its JSON object representation.
+	 *
+	 * @param jsonObjectBuilder the json object builder for the page.
+	 * @param url the URL of the collection.
+	 */
```
Comment (nhpatt) [inline vulcan-api/src/main/java/com/liferay/vulcan/message/json/PageMessageMapper.java]: Why default?

File: vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java [hunk 2]
```diff
@@ -22,17 +22,35 @@
 import javax.ws.rs.PathParam;
 
 /**
+ * Declares the endpoint from which all of your APIs originate.
+ *
+ * <p>
```
Comment (ahdezma) [inline vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java]: This one, separated by blank lines, with the html tags alone in the first and last lines

File: vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java [hunk 3]
```diff
@@ -22,17 +22,35 @@
...
  * @author Alejandro Hernández
  * @author Carlos Sierra Andrés
  * @author Jorge Ferrer
  */
 public interface RootEndpoint {
 
+	/**
+	 * Returns the single model for a given path.
+	 *
+	 * @param  path the path from the URL.
```
Comment (ahdezma) [inline vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java]: the `formatJavadoc` gradle task does that

File: vulcan-api/src/main/java/com/liferay/vulcan/message/json/PageMessageMapper.java (hunk 1 above)
Comment (ahdezma) [inline vulcan-api/src/main/java/com/liferay/vulcan/message/json/PageMessageMapper.java]: So implementations of these mappers doesn't need to implement all event, only those they want to map

File: vulcan-api/src/main/java/com/liferay/vulcan/list/FunctionalList.java [hunk 4]
```diff
@@ -19,21 +19,56 @@
 import java.util.stream.Stream;
 
 /**
+ * Instances of this class behave like a list in pure functional languages.
+ * Instead of splitting the list manually, you can use the provided methods to
+ * take different elements from the list, like <code>#head()</code>,
+ * <code>#tail()</code>, etc.
+ *
+ * Instances of <code>FunctionalList</code> should always have at least one
+ * element.
```
Comment (ahdezma) [inline vulcan-api/src/main/java/com/liferay/vulcan/list/FunctionalList.java]: Yes, this is indeed an error. I just uploaded a new commit that fix it: https://github.com/liferay/com-liferay-vulcan/pull/39/commits/c964a26227a4b3baa43c77f44dd6d210bba2a4d2

File: vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java [hunk 5]
```diff
@@ -22,17 +22,35 @@
...
 	@GET
 	@Path("/p/{path}/{id}")
 	public <T> SingleModel<T> getCollectionItemSingleModel(
 		@PathParam("path") String path, @PathParam("id") String id);
 
+	/**
+	 * Returns the collection page for a given path.
+	 *
+	 * @param  path the path from the URL.
+	 * @return the collection page at the path.
```
Comment (ahdezma) [inline vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java]: I don't understand

File: vulcan-api/src/main/java/com/liferay/vulcan/list/FunctionalList.java [hunk 6]
```diff
@@ -19,21 +19,56 @@
...
+	 *
+	 * @return a stream with all elements except the last one.
+	 */
 	public Stream<T> init();
 
+	/**
+	 * Returns the last element of the list. If the list only has one element,
+	 * this method returns the same as <code>#head()</code>
+	 *
+	 * @return the last element.
```
Comment (ahdezma) [inline vulcan-api/src/main/java/com/liferay/vulcan/list/FunctionalList.java]: Last one tail's length should be 1, while for the first one should be 0

File: vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java (hunk 2 above)
Comment (nhpatt) [inline vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java]: 👍

File: vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java (hunk 3 above)
Comment (nhpatt) [inline vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java]: 👍

File: vulcan-api/src/main/java/com/liferay/vulcan/list/FunctionalList.java (hunk 4 above)
Comment (nhpatt) [inline vulcan-api/src/main/java/com/liferay/vulcan/list/FunctionalList.java]: 👍

File: vulcan-api/src/main/java/com/liferay/vulcan/message/json/PageMessageMapper.java (hunk 1 above)
Comment (nhpatt) [inline vulcan-api/src/main/java/com/liferay/vulcan/message/json/PageMessageMapper.java]: 👍

**Comment by nhpatt** [comment]
👍

File: vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java (hunk 5 above)
Comment (ahdezma) [inline vulcan-api/src/main/java/com/liferay/vulcan/endpoint/RootEndpoint.java]: https://github.com/liferay/com-liferay-vulcan/pull/39/commits/a4769ff7c72bda2749f4747f00ebd9384f709f27

**Comment by ahdezma** [comment]
ci:retest

**Pull request merged**